src/vec_int16_ppc.h

```diff
@@ -61,9 +79,447 @@
+  vui16_t magic = vec_splats ((unsigned short) 41839);
+  const int s = 14;
+  vui16_t tmp, rem;
+
+  q = vec_mulhuh (magic, n);
+    {
+      const vui16_t vec_ones = vec_splat_u16 ( 1 );
+      vui16_t n_1 = vec_sub (n, vec_ones);
+      // avg = (q + (n-1) + 1) >> 1
+      q = vec_avg (q, n_1);
+      result = vec_srhi (q, (s - 1));
+    }
+  return result;
+}
+ * \endcode
+ *
+ * \note vec_avg(), vec_sub(), vec_splats() and vec_splat_u16() are
+ * existing altivec.h generic built-ins.
+ *
+ * The modulo computation remains the same as \ref int16_examples_0_1_1.
+ *
+ * \section int16_perf_0_0 Performance data.
+ *
+ * We can use the example above (see \ref i16_endian_issues_0_1) to
+ * illustrate the performance metrics pveclib provides.
+ * For vec_mulhuh() the core operation is the sequence
+ * vmulouh/vmuleuh/vperm. This represents the best case latency,
+ * when it is used multiple times in a single larger function.
+ *
+ * The compiler notes that vmulouh/vmuleuh are independent
+ * instructions that can execute concurrently
+ * (in separate vector pipelines).
+ * The compiler schedules them to issue in same cycle.
+ * The latency for vmulouh/vmuleuh is listed as 7 cycle and the
+ * throughput of 2 per cycle (there are 2 vector pipes for multiply).
+ * As we assume this function will use both vector pipelines, the
+ * throughput for this function is reduced to 1 per cycle.
+ *
+ * We still need to select/merge the results. The vperm instruction is
+ * dependent on the execution of both vmulouh/vmuleuh and load of the
+ * select vector complete.  For this case we assume that the load of
+ * the permute select vector has already executed.  The processor can
+ * not issue the vperm until both vmulouh/vmuleuh instructions execute.
+ * The latency for vperm is 2 cycles (3 on POWER9).  So the best case
+ * latency for this operation is is (7 + 2 = 9) cycles (10 on POWER9).
+ *
+ * Looking at the first or only execution of vec_mulhuh() in a function
+ * defines the worse case latency.
+ * Here we have to include the permute select vector load and (for LE)
+ * the select vector complement.
+ * However this case provides additional multiple pipe parallelism that
+ * needs to be accounted for in the latencies.
+ *
+ * The compiler notes that addis/vmulouh/vmuleuh are independent
+ * instructions that can execute concurrently in separate pipelines.
+ * So the compiler schedules them to issue in same cycle.
+ * The latency for vmulouh/vmuleuh is 7 cycles while the addis latency
+ * is only 2 cycles.  The dependent addi instruction can issue in
+ * the 3rd cycle, while vmulouh/vmuleuh are still executing.
+ * The addi also has a 2 cycle latency, so the dependent lvx can issue
+ * in the 5th cycle, while vmulouh/vmuleuh are still executing.
+ * The lvx has a latency of 5 cycles and will not complete execution
+ * until 2 cycles after vmulouh/vmuleuh.
+ * The dependent xxlnor is waiting of the load (lvx)
+ * and has a latency of 2 cycles.
```

Comment:
You're enjoying this, aren't you?  ;-)

Reply:
No that there is anything wrong with that ;)